### PR TITLE
Fix Crinkle Arcs

### PR DESCRIPTION
**Pull request by ptrgags**
The goal of this PR is to finish the Crinkle Arcs example.

Draft for now, I want to think about the orthogonal arc computation, as this operation will be useful for future art.

### CODE REVIEW

File: mobius/examples/crinkle_arc.rs [hunk 1]
```diff
@@ -42,30 +42,49 @@ fn get_orthog_circle(circle: Circle, a: Complex, b: Complex) -> Circle {
     }
 }
 
-fn arc_lerp(circle: Circle, a: Complex, b: Complex, t: f64) -> Complex {
+fn lerp(a: f64, b: f64, t: f64) -> f64 {
```
Comment (ptrgags) [inline mobius/examples/crinkle_arc.rs]: (In the folded code above)

I'm going to need `get_orthog_circle()` again for some future art ideas, so move this to its own file and unit test it.

Also, if the two points form a 180 degree angle, the result is the diameter of a circle, so that should return a `Line`. So this should return an enum.

🤔 There's already a `GeneralizedCircle` enum in `cline.rs`. Maybe move that to `geometry`?

File: mobius/src/geometry/circle.rs [hunk 2]
```diff
@@ -29,6 +29,10 @@ impl Circle {
     pub fn get_angle(&self, point: Complex) -> Option<f64> {
         (point - self.center).arg()
     }
+
+    pub fn point_inside(&self, point: Complex) -> bool {
```
Comment (ptrgags) [inline mobius/src/geometry/circle.rs:33]: Add unit tests for this struct.

File: mobius/src/geometry/circular_arc.rs [hunk 3]
```diff
@@ -39,3 +106,12 @@ impl Display for CircularArc {
         write!(f, "Arc(c={}, r={:.3}, {})", center, radius, angles)
     }
 }
+
+#[cfg(test)]
+mod test {
+
+    #[test]
+    pub fn missing_tests() {
+        todo!("from_circle_and_points, directed_edge");
```
Comment (ptrgags) [inline mobius/src/geometry/circular_arc.rs]: Add these tests

File: mobius/src/geometry/circular_arc.rs [hunk 4]
```diff
@@ -15,6 +35,53 @@ impl CircularArc {
...
+        }?;
+
+        Ok(Self { circle, angles })
+    }
+
+    pub fn direction(&self) -> ArcDirection {
+        self.angles.direction()
+    }
+
+    pub fn midpoint(&self) -> Complex {
```
Comment (ptrgags) [inline mobius/src/geometry/circular_arc.rs]: Instead of this, have an `interpolate(&self, t: f64)` method that computes an arbitrary fraction of the angular range, that's more general. Make sure to test.

File: mobius/examples/crinkle_arc.rs [hunk 5]
```diff
@@ -74,26 +93,26 @@ fn arc_fractal(arc: CircularArc) -> (Mobius, Mobius) {
 }
 
 fn main() -> Result<(), Error> {
-    let angles = ArcAngles::new(0.0, PI / 2.0).unwrap();
+    let angles = ArcAngles::new(0.0, 3.0 * PI / 4.0).unwrap();
```
Comment (ptrgags) [inline mobius/examples/crinkle_arc.rs]: I need to try various arcs and see what arrangement looks most artistic

File: mobius/examples/crinkle_arc.rs [hunk 6]
```diff
@@ -74,26 +93,26 @@ fn arc_fractal(arc: CircularArc) -> (Mobius, Mobius) {
...
     let tile: ClineArc = arc.into();
 
     let ifs = SemigroupIFS::new(vec![a, b]);
 
     let tiles = ifs.apply(&tile, 0, 8);
-    let orange_lines = Style::stroke(255, 127, 0).with_width(0.5);
+    let orange_lines = Style::stroke(255, 127, 0).with_width(0.125);
     let geometry = style_geometry(orange_lines, &tiles[..]);
 
     let orthog_circle = get_orthog_circle(Circle::unit_circle(), Complex::ONE, Complex::I);
```
Comment (ptrgags) [inline mobius/examples/crinkle_arc.rs]: The boundary of the whole image is a lens between the original arc and the inside arc of this orthogonal circle. The two transformations transform this lens into two smaller lenses. This leaves an ideal triangle gap formed by the orthogonal arc, and the two sub-arcs of the first arc.

It would be helpful to compute both of these shapes. Iterate the lens a number of times but only render the leaves. Meanwhile, iterate the triangle and render every iteration.

I think this might look best if those shapes are filled in two different colors, though I don't yet have polygon rendering. For now, just compute the shapes and stroke the results.

Comment (ptrgags) [reply mobius/examples/crinkle_arc.rs]: Also play with the iteration depth. Yes, it's a fractal, but the limit will be a crinkle-y line. A chain of lenses would look more interesting.

File: mobius/examples/crinkle_arc.rs [hunk 7]
```diff
@@ -42,30 +42,49 @@ fn get_orthog_circle(circle: Circle, a: Complex, b: Complex) -> Circle {
...
-    circle.get_point(lerp_angle)
+    if original_circle.point_inside(mid1) {
+        CircularArc::from_circle_and_points(circle, a, mid1, b).unwrap()
+    } else {
+        CircularArc::from_circle_and_points(circle, a, mid2, b).unwrap()
+    }
 }
 
-fn arc_fractal(arc: CircularArc) -> (Mobius, Mobius) {
+fn arc_fractal(arc: CircularArc, t: f64) -> (Mobius, Mobius) {
```
Comment (ptrgags) [inline mobius/examples/crinkle_arc.rs]: This is close, but here's a better way to phrase the algorithm:

- start with arc `ac`
- interpolate to get point `b` in between a and c at `t` percent of the way across. So the arc is `abc`
- Compute two orthogonal arcs (always same orientation as the original circle, and always inside the first circle). These sub arcs are `cb` and `ba`
- interpolate points 't' percent across each arc, so now we have `ceb` and `bda`
- Compute two transformations by sending `abc -> ceb` and `abc -> bda`

Think about how to clean up the code to phrase it like the above.

File: mobius/examples/crinkle_arc.rs [hunk 8]
```diff
@@ -42,30 +42,49 @@ fn get_orthog_circle(circle: Circle, a: Complex, b: Complex) -> Circle {
...
     let CircularArc { circle, angles } = arc;
     let ArcAngles(angle_a, angle_c) = angles;
-    let t = 0.5;
 
     let a = circle.get_point(angle_a);
-    let angle_b = t * (angle_a + angle_c);
+    let angle_b = lerp(angle_a, angle_c, t);
     let b = circle.get_point(angle_b);
     let c = circle.get_point(angle_c);
```
Comment (ptrgags) [inline mobius/examples/crinkle_arc.rs]: This could be written more expressively

- `a` is just `arc.start()`
` `c` is `arc.end()`
` `b` is `arc.interpolate(t)`

File: mobius/examples/crinkle_arc.rs [hunk 9]
```diff
@@ -42,30 +42,49 @@ fn get_orthog_circle(circle: Circle, a: Complex, b: Complex) -> Circle {
...
 
     let circle_ab = get_orthog_circle(circle, a, b);
     let circle_bc = get_orthog_circle(circle, b, c);
 
-    let d = arc_lerp(circle_ab, b, a, t);
-    let e = arc_lerp(circle_bc, c, b, t);
+    let arc_ab = compute_sub_arc(circle, circle_ab, a, b, t);
+    let arc_bc = compute_sub_arc(circle, circle_bc, b, c, t);
```
Comment (ptrgags) [inline mobius/examples/crinkle_arc.rs]: While the calculations will be the same, conceptually, the operation is `compute_orthogonal_arc(arc, a, b, t)`

For now, have this method compute the _inside_ portion of the arc, and have it go in the same direction as the original arc. In the future, I may want to have flags for this.

Comment (ptrgags) [reply mobius/examples/crinkle_arc.rs]: Though also, even without the flags, `complement()` and `reverse()` get you the rest of the way.

**Review by ptrgags — COMMENTED** [review]
Looks like there's still quite a bit to do.

File: mobius/src/geometry/circular_arc.rs [hunk 10]
```diff
@@ -15,6 +35,59 @@ impl CircularArc {
...
+        // Get the raw angles
+        let theta_a = circle.get_angle(a).unwrap();
+        let theta_b = circle.get_angle(b).unwrap();
+        let theta_c = circle.get_angle(c).unwrap();
+
+        let angles = if ccw {
+            compute_ccw_angles(theta_a, theta_b, theta_c)
+        } else {
+            compute_cw_angles(theta_a, theta_b, theta_c)
+        }?;
```
Comment (ptrgags) [inline mobius/src/geometry/circular_arc.rs]: I think these could be replaced by calls to `ArcAngles::from_raw_angles(a, b, CounterClockwise|Clockwise)`

File: mobius/src/geometry/line.rs [hunk 11]
```diff
@@ -11,14 +50,40 @@ pub struct Line {
...
-    pub fn new(normal: Complex, distance: f64) -> Result<Self, String> {
-        match normal.normalize() {
-            Some(unit_normal) => Ok(Self {
-                unit_normal,
-                distance,
-            }),
-            None => Err(String::from("Normal must be finite and non-zero")),
+    pub fn new(normal: Complex, distance: f64) -> Result<Self, LineError> {
+        ComplexError::require_finite_nonzero("normal", normal)?;
+        FloatError::require_finite("distance", distance)?;
```
Comment (ptrgags) [inline mobius/src/geometry/line.rs:55]: I don't like adding these checks all over the place, it feels more like validation code.

It might be nice to have `FiniteNonzero` or `Finite` wrapper around a Complex to restrict the types... though not sure I like the extra nesting and boilerplate required.

File: mobius/src/interpolation.rs [hunk 12]
```diff
@@ -0,0 +1,3 @@
+pub fn lerp(a: f64, b: f64, t: f64) -> f64 {
```
Comment (ptrgags) [inline mobius/src/interpolation.rs:1]: Add tests for this, since I'd imagine this will be used a lot

File: mobius/src/geometry/line.rs [hunk 13]
```diff
@@ -11,14 +50,40 @@ pub struct Line {
...
     /// Create a line from a normal and distance offset
     /// This will automatically normalize the normal
-    pub fn new(normal: Complex, distance: f64) -> Result<Self, String> {
-        match normal.normalize() {
-            Some(unit_normal) => Ok(Self {
-                unit_normal,
-                distance,
-            }),
-            None => Err(String::from("Normal must be finite and non-zero")),
+    pub fn new(normal: Complex, distance: f64) -> Result<Self, LineError> {
```
Comment (ptrgags) [inline mobius/src/geometry/line.rs:53]: I don't like the way this constructor works, because it only normalizes the normal, not the distance value as well. rethink this..

File: mobius/src/geometry/circle.rs [hunk 14]
```diff
@@ -29,6 +29,16 @@ impl Circle {
...
+        (point - self.center).norm() <= self.radius * self.radius
+    }
+}
+
+impl PartialEq for Circle {
```
Comment (ptrgags) [inline mobius/src/geometry/circle.rs:38]: Add some tests for this.

File: mobius/examples/crinkle_arc.rs [hunk 15]
```diff
@@ -1,99 +1,109 @@
...
+            arc,
+            orthog_arc: compute_orthogonal_arc(arc, a, c),
+            sub_arcs: (arc_cb, arc_ba),
+        }
+    }
 }
 
 fn main() -> Result<(), Error> {
-    let angles = ArcAngles::new(0.0, PI / 2.0).unwrap();
+    let angles = ArcAngles::new(-PI / 2.0, PI / 4.0).unwrap();
```
Comment (ptrgags) [inline mobius/examples/crinkle_arc.rs]: This didn't work when I used an angle greater than 180 degrees. Investigate.

File: mobius/examples/crinkle_arc.rs [hunk 16]
```diff
@@ -1,99 +1,109 @@
...
-    let angle_a = circle.get_angle(a).unwrap();
-    let angle_b = circle.get_angle(b).unwrap();
-
-    let angle_bisector = 0.5 * (angle_a + angle_b);
-    let angle_bisector = if (angle_b - angle_a) % TAU > PI {
-        (angle_bisector + PI) % TAU
-    } else {
-        angle_bisector
+
+fn compute_orthogonal_arc(arc: CircularArc, a: Complex, b: Complex) -> CircularArc {
```
Comment (ptrgags) [inline mobius/examples/crinkle_arc.rs]: This should be moved to the new file and tested. Especially since large arcs seem to cause it to fall apart.

File: mobius/examples/crinkle_arc.rs (hunk 6 above)
Comment (ptrgags) [reply mobius/examples/crinkle_arc.rs]: Actually, I went back to my original sketch on paper. It looks neat when you start with the triangles, and alternate colors every iteration. 

This could be done by taking the original triangle and the IFS `{aa, ab, ba, bb}` in the first color.

Then in the second triangle:
- take `a(triangle)` and iterate with `{aa, ab}`
- take `b(triangle)` and iterate with `{ba, bb}`

File: mobius/src/geometry/line.rs (hunk 11 above)
Comment (ptrgags) [reply mobius/src/geometry/line.rs:55]: I'm still chewing on this idea, but it's definitely out of scope.

File: mobius/src/geometry/line.rs (hunk 13 above)
Comment (ptrgags) [reply mobius/src/geometry/line.rs:53]: Thinking about it, if you divide the normal by its magnitude, the distance has to be divided by the same amount. Add a unit test and implementation for this.

File: mobius/examples/crinkle_arc.rs (hunk 15 above)
Comment (ptrgags) [reply mobius/examples/crinkle_arc.rs]: It's because I'm assuming the inside arc is in the same direction as the original arc. That's not true if the original arc is > pi.

`ArcAngles` should have a `get_central_angle` which computes `|b - a|`

File: mobius/examples/crinkle_arc.rs [hunk 17]
```diff
@@ -1,99 +1,109 @@
...
+        GeneralizedCircle::Line(_) => panic!("Not implemented: sub arc that's a line"),
     };
 
-    let orthog_center = Complex::from_polar(p, angle_bisector);
+    // My convention is to compute the sub arc that's sweeping in the same
+    // angular direction as the original arc. But if the original one went from b -> a,
+    // now we're going from a -> b;
+    let angle_a_raw = orthog_circle.get_angle(a).unwrap();
+    let angle_b_raw = orthog_circle.get_angle(b).unwrap();
+    let sub_angles = ArcAngles::from_raw_angles(angle_b_raw, angle_a_raw, arc.direction());
```
Comment (ptrgags) [inline mobius/examples/crinkle_arc.rs]: If the central angle is greater than $\pi$, we want the complement instead.

File: mobius/examples/crinkle_arc.rs [hunk 18]
```diff
@@ -1,99 +1,109 @@
...
-    let group = Group::new().add(geometry).add(more_geometry);
-
-    render_views("output", "crinkle_arc", &[View("", 0.5, 0.5, 0.51)], group)?;
+    let lens_tile = ClineArcTile::new(vec![fractal.arc.into(), fractal.orthog_arc.into()]);
+
+    let (arc_cb, arc_ba) = fractal.sub_arcs;
+    let triangle_tile = ClineArcTile::new(vec![
+        // orthogonal arc from b -> a
+        fractal.orthog_arc.into(),
+        //
```
Comment (ptrgags) [inline mobius/examples/crinkle_arc.rs]: finish commenting here.

File: mobius/examples/crinkle_arc.rs [hunk 19]
```diff
@@ -1,99 +1,109 @@
...
+        arc_ba.reverse().into(),
+        arc_cb.reverse().into(),
+    ]);
+
+    let depth = 7usize;
+
+    let triangle_tiles = ifs.apply(&triangle_tile, 0, depth - 1);
+    let leaf_lenses = ifs.apply(&lens_tile, depth, depth);
```
Comment (ptrgags) [inline mobius/examples/crinkle_arc.rs:84]: This iteration will look a little different to do the two-color method.

File: mobius/examples/crinkle_arc.rs [hunk 20]
```diff
@@ -1,99 +1,109 @@
...
+        Self {
+            xforms: (xform_ceb, xform_bda),
+            arc,
+            orthog_arc: compute_orthogonal_arc(arc, a, c),
+            sub_arcs: (arc_cb, arc_ba),
+        }
+    }
 }
 
 fn main() -> Result<(), Error> {
```
Comment (ptrgags) [inline mobius/examples/crinkle_arc.rs]: Once I get this working for one arc, make a pattern that chains several arcs together into a big crinkly necklace.

File: mobius/src/geometry/arc_angles.rs [hunk 21]
```diff
@@ -251,4 +281,27 @@ mod test {
...
+
+        let rev_comp = arc.complement().reverse();
+        let comp_rev = arc.reverse().complement();
+
+        assert_eq!(rev_comp, comp_rev);
+    }
+
+    #[test]
+    pub fn missing_tests() {
+        todo!("test from_raw_angles")
```
Comment (ptrgags) [inline mobius/src/geometry/arc_angles.rs]: Don't forget to add tests here!

File: mobius/examples/crinkle_arc.rs [hunk 22]
```diff
@@ -1,99 +1,112 @@
...
-    let angles = ArcAngles::new(0.0, PI / 2.0).unwrap();
+    let angles = ArcAngles::new(-PI / 2.0, PI).unwrap();
```
Comment (ptrgags) [inline mobius/examples/crinkle_arc.rs]: This works now... but if the arc crosses the atan2 branch point of +/- PI, things go wrong.

File: mobius/src/geometry/circle.rs (hunk 2 above)
Comment (ptrgags) [reply mobius/src/geometry/circle.rs:33]: Actually, I want to come back to this. I have some ideas for wrapper types that restrict the possible values, let me add those before testing

File: mobius/src/orthogonal_arcs.rs [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+use crate::{
+    geometry::{ArcAngles, Circle, CircularArc, GeneralizedCircle, Line, LineSegment},
+    nearly::is_nearly,
+    Complex,
+};
+
+#[derive(Debug)]
+pub enum OrthogonalArcError {
+    DuplicatePoint(Complex),
+    PointAtCenter(Complex),
```
Comment (ptrgags) [inline mobius/src/orthogonal_arcs.rs]: Looking at this again, I don't like that it's possible to pass in a point not on the original circle. Specifying an ArcAngles would be cleaner I think

File: mobius/src/orthogonal_arcs.rs [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+    let orthog_center = circle.center + Complex::from_polar(p, angle_bisector);
+    let orthog_circle = Circle {
+        center: orthog_center,
+        radius: orthog_radius,
+    };
+
+    Ok(GeneralizedCircle::Circle(orthog_circle))
+}
+
+pub fn compute_orthogonal_arc(arc: CircularArc, a: Complex, b: Complex) -> CircularArc {
```
Comment (ptrgags) [inline mobius/src/orthogonal_arcs.rs]: Similarly here. The arc already specifies angles, get the points from there.

If crinkle_arc needs to adjust the angles to call this method, I think that's okay.

Comment (ptrgags) [reply mobius/src/orthogonal_arcs.rs]: This worked out nicely!

File: mobius/src/geometry/circular_arc.rs [hunk 25]
```diff
@@ -15,6 +42,81 @@ impl CircularArc {
     pub fn new(circle: Circle, angles: ArcAngles) -> Self {
         Self { circle, angles }
     }
+
+    /// Given a circle and 3 points (a, b, c) on that circle, compute a circular
+    /// arc from a -> b -> c
+    pub fn from_circle_and_points(
```
Comment (ptrgags) [inline mobius/src/geometry/circular_arc.rs]: I regret moving this into `CircularArc`. Too many corner cases to handle.

a `ClineArc` stores 3 points that are guaranteed to be on the circle (in the `From::<CircularArc>()` constructor). But this constructor here does not have that guarantee.

Also note that `ClineArc` is the only struct that uses this method.

So I think it should be moved back into `ClineArc`, then this struct becomes simpler and there are less corner cases to test.

File: mobius/src/interpolation.rs (hunk 12 above)
Comment (ptrgags) [reply mobius/src/interpolation.rs:1]: Let's see how much it gets used in practice. For now, it's an implementation detail

File: mobius/src/geometry/circle.rs (hunk 14 above)
Comment (ptrgags) [reply mobius/src/geometry/circle.rs:38]: I want to revisit constraints on all geometry primitives, I'll circle back to this later.

File: mobius/examples/crinkle_arc.rs (hunk 6 above)
Comment (ptrgags) [reply mobius/examples/crinkle_arc.rs]: This worked out nicely!
